[ADD] [16.0] partner_resource_delivery_schedule

Shide opened the pull request:
Allows to define a Calendar (and a resource) on a res.partner

This resource and calendar will be used to offer a better expected_date than Odoo does.
It takes in consideration Company calendar and Partner Resource and Calendar with his leaves too.

PR Series: https://github.com/OCA/delivery-carrier/pull/748

MT-4311 @moduon @yajo @EmilioPascual @rafaelbn please review if you want :)

Shide commented:
This module `delivery_partner_schedule` tests are not failing, but other modules do.

Shide committed:
[ADD] partner_resource_delivery_schedule

EmilioPascual commented:
Thanks for this modules @Shide 

When I try to test it, I couldn't select or create any avaliability resource

![image](https://github.com/OCA/delivery-carrier/assets/53056345/47e0a656-d118-4ab6-9227-88afb012cfc3)

EmilioPascual commented on delivery_partner_schedule/data/ir_config_parameter.xml:
```diff
@@ -0,0 +1,7 @@
+<?xml version="1.0" encoding="UTF-8" ?>
+<data noupdate="1">
+    <record id="interval_days" model="ir.config_parameter">
```
Perhaps you should report this behaviour in the readme

Shide commented:
> Thanks for this modules @Shide 
> 
> When I try to test it, I couldn't select or create any avaliability resource
> 
> ![image](https://github.com/OCA/delivery-carrier/assets/53056345/47e0a656-d118-4ab6-9227-88afb012cfc3)
> 

Start typing and you will see the create and edit options

EmilioPascual commented:
> > Thanks for this modules @Shide
> > When I try to test it, I couldn't select or create any avaliability resource
> > ![image](https://private-user-images.githubusercontent.com/53056345/293655029-47e0a656-d118-4ab6-9227-88afb012cfc3.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDQyNzc0NTgsIm5iZiI6MTcwNDI3NzE1OCwicGF0aCI6Ii81MzA1NjM0NS8yOTM2NTUwMjktNDdlMGE2NTYtZDExOC00YWI2LTkyMjctODhhZmIwMTJjZmMzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMDMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTAzVDEwMTkxOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWExZGUzODAwNDhhNTdmNDUxZTU2NDU1ZDYyYjU2MWQ5YmFlNzg2ZjM4NDIzZTY5MWU1YWJlZGU0MWU5MDM2MmMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.96VzKIGQKs_9BPdfw3W7V6iRoiJTWdfJv23Yw3_oJKw)
> 
> Start typing and you will see the create and edit options

Yes I tried, but I was not able to create a resource. I show you in following video:
https://www.loom.com/share/7810f26218e24d6ca2798018e8e76fe5

Shide commented:
@EmilioPascual you can create it as you will expect

Shide commented:
Module renamed

EmilioPascual commented on partner_resource_delivery_schedule/readme/USAGE.md:
```diff
@@ -0,0 +1,24 @@
+
+1. Go to a Customer
+2. Create an Availability Resource under Sales page
+3. Create a Sales Order for this Customer with a non Service product with Quantity > 1
```
```suggestion
3. Create a Sales Order for this Customer with a non Service product with Quantity > 0
```

EmilioPascual commented on partner_resource_delivery_schedule/models/res_partner.py:
```diff
@@ -0,0 +1,15 @@
+        string="Availability Resource",
+        domain="[('resource_type', '=', 'contact')]",
+        required=False,
+    )
```
```suggestion
    )
```
Maybe you can add a M2O to resource.calendar, when the calendar resource changes, update the resource. This way it is easier for the user to modify the available resource like Odoo does on the employee.

Shide commented on partner_resource_delivery_schedule/models/res_partner.py:
```diff
@@ -0,0 +1,15 @@
+
+class ResPartner(models.Model):
+    _name = "res.partner"
+    _inherit = ["res.partner", "resource.mixin"]
```
Remove resource.mixing because when creating a contact, needs to have permisions for the resource manager

Shide commented:
I've changed a bit the module to simplify process and improve UX:
- A new button appears to: create a resource and show a calendar.
- When calendar changes, it's written too in the resource of the contact

https://www.loom.com/share/19005ee119b046758340f7c4b344afb6

Shide commented:
@rafaelbn @fcvalgar @Gelojr @EmilioPascual plz review if you want :)

fcvalgar reviewed (approved):
Functional review

LGTM, Thank you @Shide 


![2024-01-23_16-33](https://github.com/OCA/delivery-carrier/assets/17847432/24bb2efc-5cfe-4919-aeb2-0641e4ffcd89)
![2024-01-23_16-33_1](https://github.com/OCA/delivery-carrier/assets/17847432/b0d8bf7c-2547-44f2-ab56-1ecdc8979fd3)
![image](https://github.com/OCA/delivery-carrier/assets/17847432/de0aea2d-acca-4ed7-adc3-b1112a57212f)

yajo commented on partner_resource_delivery_schedule/views/res_partner_view.xml:
```diff
@@ -0,0 +1,35 @@
+            <xpath expr="//group[@name='sale']" position="inside">
+                <field
+                    name="delivery_schedule_resource_id"
+                    domain="[('resource_type', '=', 'delivery')]"
```
**suggestion:** put the domain in the model instead.

yajo commented on partner_resource_delivery_schedule/readme/CONTEXT.md:
```diff
@@ -0,0 +1,9 @@
+
+If you need this module for those reasons, these might interest you too:
+
+- delivery_carrier_resource_schedule
```
**suggestion:** avoid markdown doing weird things with those `_`

```suggestion
- `delivery_carrier_resource_schedule`
```

yajo commented on partner_resource_delivery_schedule/readme/USAGE.md:
```diff
@@ -0,0 +1,26 @@
+
+Configuration:
+
+1. Configure ``partner_resource_delivery_schedule.interval_days`` parameter and set
```
**issue:** this is markdown now! (finally) 😄 

```suggestion
1. Configure `partner_resource_delivery_schedule.interval_days` parameter and set
```

yajo reviewed (approved):
**praise:** Some suggestions, nothing blocking. Good one!

EmilioPascual reviewed (changes requested):
One user with sales administrator group cannot create a delivery schedule calendar.
I show you in the following recording https://www.loom.com/share/57a81d467a274252a35f08423800fdb2
Any user with sales group should be able to create a delivery schedule calendar.

rousseldenis commented on partner_resource_delivery_schedule/models/sale_order_line.py:
```diff
@@ -0,0 +1,19 @@
+        values = super()._prepare_procurement_values(group_id)
+        if self.order_id.commitment_date:
+            return values
+        values["date_deadline"] = self.order_id.expected_date
```
I'm wondering if these shouldn't go in another module...

Shide replied:
These lines are present in `sale_stock` module and this module depends on it.
https://github.com/odoo/odoo/blob/16.0/addons/sale_stock/models/sale_order_line.py#L233

rousseldenis commented on partner_resource_delivery_schedule/tests/test_partner_expected_date.py:
```diff
@@ -0,0 +1,161 @@
+@tagged("post_install", "-at_install")
+class TestPartnerExpectedDate(TransactionCase):
+    def setUp(self):
+        super(TestPartnerExpectedDate, self).setUp()
```
@Shide Could you use classmethod instead ?

FYI, to help performances, it is now better to inherit from https://github.com/odoo/odoo/blob/16.0/odoo/addons/base/tests/common.py#L19

Shide replied:
Using Transactioncase from sale module was an error!

Shide replied:
I want to use SetUp instead setUpClass because I'm doing an assertion when creating the resource.calendar:
https://github.com/OCA/delivery-carrier/pull/747/files?diff=unified&w=0#diff-db3a6b820ce4edee36af473c6f5c0039a4c82c8af235241eab74c6f79ba9eb40R51-R55

`cls` doesn't have assert methods.

Shide commented:
Tests are failing on other module.

yajo commented:
Try setting the module that fails (and its dependencies, if any) as rebel by using the template.

Shide commented:
I've changed the way that the available date is set on the sales order.

I've added a small button (>>) to suggest the next available date on the commitment field. 
If you click again the button, a new date is calculated from the prior one.

This behavior respects Odoo expected date in case you won't need to promise anything to the customer.

@rafaelbn @rousseldenis what do you think?

Gelojr reviewed (changes requested):
Hello @Shide 
It seems that when you create a new client, the selection field does not work until you start typing and create a new calendar.
But when you do, the rest of the calendars do not appear. Is it OK?
https://www.loom.com/share/7718a4c33dc3484495f2d0a3fa2059ec?sid=6ec590e3-90df-4f7a-80a9-61199f691dda

Shide commented:
> Hello @Shide It seems that when you create a new client, the selection field does not work until you start typing and create a new calendar. But when you do, the rest of the calendars do not appear. Is it OK? https://www.loom.com/share/7718a4c33dc3484495f2d0a3fa2059ec?sid=6ec590e3-90df-4f7a-80a9-61199f691dda

Yes, because by default, the company calendar is assigned.
When you want to choose another calendar, the domain is applied.

rafaelbn reviewed (changes requested):
I don't feel it. I promise A deep review, with love 💕❤️

Shide commented:
I'm going to think about this PR and rebuild the solution again with something less generic (resource.calendar) that uses non general permissions.

rousseldenis commented:
> I'm going to think about this PR and rebuild the solution again with something less generic (resource.calendar) that uses non general permissions.

@Shide FYI, there is already something : https://github.com/OCA/stock-logistics-workflow/pull/1378

Shide commented:
> > I'm going to think about this PR and rebuild the solution again with something less generic (resource.calendar) that uses non general permissions.
> 
> @Shide FYI, there is already something : https://github.com/OCA/stock-logistics-workflow/pull/1378

We know, but this solution is not what we are looking for
